fedora 40 rust crate compile error

```
cargo new kuzu-test
cd kuzu-test
cargo add kuzu
cargo build
```

gcc version: gcc-14.1.1-6

log:
```
-- Build files have been written to: /home/oleh/Work/Rust/rust_kuzu_test/target/debug/build/kuzu-aedabeca820bb46a/out/build
  running: cd "/home/oleh/Work/Rust/rust_kuzu_test/target/debug/build/kuzu-aedabeca820bb46a/out/build" && MAKEFLAGS="-j --jobserver-fds=8,9 --jobserver-auth=8,9" "cmake" "--build" "." "--config" "Debug"
  [  0%] Building C object third_party/serd/CMakeFiles/serd.dir/src/base64.c.o
  [  0%] Building C object third_party/brotli/CMakeFiles/brotlicommon.dir/c/common/constants.c.o
  [  0%] Building CXX object third_party/fastpfor/CMakeFiles/fastpfor.dir/fastpfor/bitpacking.cpp.o
  [  0%] Building CXX object third_party/re2/CMakeFiles/re2.dir/bitstate.cpp.o
  [  1%] Building CXX object third_party/miniparquet/CMakeFiles/miniparquet.dir/src/parquet/parquet_constants.cpp.o
  [  1%] Building CXX object third_party/miniz/CMakeFiles/miniz.dir/miniz.cpp.o
  [  2%] Building CXX object third_party/mbedtls/CMakeFiles/mbedtls.dir/library/aes.cpp.o
  [  2%] Building CXX object third_party/antlr4_runtime/CMakeFiles/antlr4_runtime.dir/src/ANTLRErrorListener.cpp.o
  [  2%] Building C object third_party/brotli/CMakeFiles/brotlicommon.dir/c/common/context.c.o
  [  3%] Building C object third_party/brotli/CMakeFiles/brotlicommon.dir/c/common/dictionary.c.o
  [  3%] Building C object third_party/serd/CMakeFiles/serd.dir/src/byte_source.c.o
  [  3%] Building C object third_party/serd/CMakeFiles/serd.dir/src/env.c.o
  [  3%] Building CXX object third_party/mbedtls/CMakeFiles/mbedtls.dir/library/aria.cpp.o
  [  3%] Building C object third_party/serd/CMakeFiles/serd.dir/src/n3.c.o
  [  3%] Building CXX object third_party/mbedtls/CMakeFiles/mbedtls.dir/library/asn1parse.cpp.o
  [  3%] Building C object third_party/brotli/CMakeFiles/brotlicommon.dir/c/common/platform.c.o
  [  3%] Building CXX object third_party/mbedtls/CMakeFiles/mbedtls.dir/library/base64.cpp.o
  [  3%] Building C object third_party/brotli/CMakeFiles/brotlicommon.dir/c/common/shared_dictionary.c.o
  [  3%] Building CXX object third_party/mbedtls/CMakeFiles/mbedtls.dir/library/bignum.cpp.o
  [  4%] Building C object third_party/brotli/CMakeFiles/brotlicommon.dir/c/common/transform.c.o
  [  4%] Building C object third_party/serd/CMakeFiles/serd.dir/src/node.c.o
  [  4%] Linking C static library libbrotlicommon.a
  [  4%] Built target brotlicommon
  [  4%] Building C object third_party/serd/CMakeFiles/serd.dir/src/reader.c.o
  [  4%] Building CXX object third_party/miniparquet/CMakeFiles/miniparquet.dir/src/parquet/parquet_types.cpp.o
  [  4%] Building CXX object third_party/mbedtls/CMakeFiles/mbedtls.dir/library/camellia.cpp.o
  [  4%] Building C object third_party/serd/CMakeFiles/serd.dir/src/string.c.o
  [  4%] Building C object third_party/serd/CMakeFiles/serd.dir/src/system.c.o
  [  4%] Building CXX object third_party/mbedtls/CMakeFiles/mbedtls.dir/library/cipher.cpp.o
  [  4%] Building C object third_party/serd/CMakeFiles/serd.dir/src/uri.c.o
  [  5%] Linking CXX static library libminiz.a
  [  5%] Building CXX object third_party/mbedtls/CMakeFiles/mbedtls.dir/library/cipher_wrap.cpp.o
  [  5%] Built target miniz
  [  5%] Building C object third_party/serd/CMakeFiles/serd.dir/src/writer.c.o
  [  5%] Building CXX object third_party/antlr4_runtime/CMakeFiles/antlr4_runtime.dir/src/ANTLRErrorStrategy.cpp.o
  [  6%] Building CXX object third_party/mbedtls/CMakeFiles/mbedtls.dir/library/constant_time.cpp.o
  [  6%] Building CXX object third_party/mbedtls/CMakeFiles/mbedtls.dir/library/entropy.cpp.o
  [  6%] Building CXX object third_party/mbedtls/CMakeFiles/mbedtls.dir/library/entropy_poll.cpp.o
  [  7%] Linking C static library libserd.a
  [  7%] Building CXX object third_party/mbedtls/CMakeFiles/mbedtls.dir/library/gcm.cpp.o
  [  7%] Building CXX object third_party/re2/CMakeFiles/re2.dir/compile.cpp.o
  [  7%] Built target serd
  [  7%] Building CXX object third_party/antlr4_runtime/CMakeFiles/antlr4_runtime.dir/src/ANTLRFileStream.cpp.o
  [  7%] Building CXX object third_party/mbedtls/CMakeFiles/mbedtls.dir/library/md.cpp.o
  [  7%] Building CXX object third_party/antlr4_runtime/CMakeFiles/antlr4_runtime.dir/src/ANTLRInputStream.cpp.o
  [  7%] Building CXX object third_party/miniparquet/CMakeFiles/miniparquet.dir/src/thrift/protocol/TProtocol.cpp.o
  [  7%] Building CXX object third_party/mbedtls/CMakeFiles/mbedtls.dir/library/oid.cpp.o
  [  7%] Building CXX object third_party/mbedtls/CMakeFiles/mbedtls.dir/library/pem.cpp.o
  [  7%] Building CXX object third_party/mbedtls/CMakeFiles/mbedtls.dir/library/pk.cpp.o
  [  7%] Building CXX object third_party/mbedtls/CMakeFiles/mbedtls.dir/library/pk_wrap.cpp.o
  [  8%] Building CXX object third_party/mbedtls/CMakeFiles/mbedtls.dir/library/pkparse.cpp.o
  [  8%] Building CXX object third_party/mbedtls/CMakeFiles/mbedtls.dir/library/platform_util.cpp.o
  [  8%] Building CXX object third_party/mbedtls/CMakeFiles/mbedtls.dir/library/rsa.cpp.o
  [  8%] Building CXX object third_party/mbedtls/CMakeFiles/mbedtls.dir/library/rsa_alt_helpers.cpp.o
  [  8%] Building CXX object third_party/mbedtls/CMakeFiles/mbedtls.dir/library/sha1.cpp.o
  [  8%] Building CXX object third_party/mbedtls/CMakeFiles/mbedtls.dir/library/sha256.cpp.o
  [  8%] Building CXX object third_party/miniparquet/CMakeFiles/miniparquet.dir/src/thrift/transport/TTransportException.cpp.o
  [  8%] Building CXX object third_party/mbedtls/CMakeFiles/mbedtls.dir/library/sha512.cpp.o
  [  8%] Linking CXX static library libmbedtls.a
  [  8%] Building CXX object third_party/utf8proc/CMakeFiles/utf8proc.dir/utf8proc.cpp.o
  [  8%] Built target mbedtls
  [  8%] Building CXX object third_party/utf8proc/CMakeFiles/utf8proc.dir/utf8proc_wrapper.cpp.o
  [  8%] Building CXX object third_party/re2/CMakeFiles/re2.dir/dfa.cpp.o
  [  8%] Building CXX object third_party/re2/CMakeFiles/re2.dir/filtered_re2.cpp.o
  [  8%] Building CXX object third_party/zstd/CMakeFiles/zstd.dir/common/entropy_common.cpp.o
  [  8%] Building CXX object third_party/antlr4_runtime/CMakeFiles/antlr4_runtime.dir/src/atn/ActionTransition.cpp.o
  [  8%] Building CXX object third_party/zstd/CMakeFiles/zstd.dir/common/error_private.cpp.o
  [  8%] Building CXX object third_party/zstd/CMakeFiles/zstd.dir/common/fse_decompress.cpp.o
  [  8%] Building CXX object src/extension/CMakeFiles/kuzu_extension.dir/extension.cpp.o
  [  8%] Building CXX object third_party/zstd/CMakeFiles/zstd.dir/common/xxhash.cpp.o
  [  8%] Linking CXX static library libutf8proc.a
  [  8%] Built target utf8proc
  [  9%] Building CXX object third_party/zstd/CMakeFiles/zstd.dir/common/zstd_common.cpp.o
  [  9%] Building CXX object src/binder/CMakeFiles/kuzu_binder.dir/binder.cpp.o
  [  9%] Building CXX object third_party/zstd/CMakeFiles/zstd.dir/compress/fse_compress.cpp.o
  [  9%] Building CXX object third_party/zstd/CMakeFiles/zstd.dir/compress/hist.cpp.o
  [  9%] Building CXX object third_party/zstd/CMakeFiles/zstd.dir/compress/huf_compress.cpp.o
  [  9%] Building CXX object third_party/zstd/CMakeFiles/zstd.dir/compress/zstd_compress.cpp.o
  [  9%] Building CXX object src/binder/bind/CMakeFiles/kuzu_binder_bind.dir/bind_attach_database.cpp.o
  [ 10%] Building CXX object third_party/antlr4_runtime/CMakeFiles/antlr4_runtime.dir/src/atn/AmbiguityInfo.cpp.o
  [ 10%] Building CXX object third_party/zstd/CMakeFiles/zstd.dir/compress/zstd_compress_literals.cpp.o
  [ 10%] Building CXX object third_party/zstd/CMakeFiles/zstd.dir/compress/zstd_compress_sequences.cpp.o
  [ 10%] Building CXX object third_party/re2/CMakeFiles/re2.dir/mimics_pcre.cpp.o
  [ 10%] Building CXX object third_party/zstd/CMakeFiles/zstd.dir/compress/zstd_compress_superblock.cpp.o
  [ 10%] Building CXX object third_party/zstd/CMakeFiles/zstd.dir/compress/zstd_double_fast.cpp.o
  [ 11%] Building CXX object third_party/zstd/CMakeFiles/zstd.dir/compress/zstd_fast.cpp.o
  [ 11%] Building CXX object third_party/re2/CMakeFiles/re2.dir/nfa.cpp.o
  [ 11%] Building CXX object third_party/zstd/CMakeFiles/zstd.dir/compress/zstd_lazy.cpp.o
  [ 11%] Building CXX object third_party/miniparquet/CMakeFiles/miniparquet.dir/src/thrift/transport/TBufferTransports.cpp.o
  [ 11%] Building CXX object third_party/antlr4_runtime/CMakeFiles/antlr4_runtime.dir/src/atn/ArrayPredictionContext.cpp.o
  [ 11%] Building CXX object third_party/zstd/CMakeFiles/zstd.dir/compress/zstd_ldm.cpp.o
  [ 11%] Building CXX object third_party/zstd/CMakeFiles/zstd.dir/compress/zstd_opt.cpp.o
  [ 11%] Building CXX object third_party/miniparquet/CMakeFiles/miniparquet.dir/src/snappy/snappy.cc.o
  [ 11%] Building CXX object third_party/re2/CMakeFiles/re2.dir/onepass.cpp.o
  [ 11%] Building CXX object third_party/zstd/CMakeFiles/zstd.dir/decompress/huf_decompress.cpp.o
  [ 11%] Building CXX object third_party/antlr4_runtime/CMakeFiles/antlr4_runtime.dir/src/atn/ATNConfig.cpp.o
  [ 11%] Building CXX object third_party/miniparquet/CMakeFiles/miniparquet.dir/src/snappy/snappy-sinksource.cc.o
  [ 11%] Linking CXX static library libminiparquet.a
  [ 11%] Building CXX object third_party/zstd/CMakeFiles/zstd.dir/decompress/zstd_ddict.cpp.o
  [ 11%] Built target miniparquet
  [ 11%] Building CXX object src/binder/bind/CMakeFiles/kuzu_binder_bind.dir/bind_comment_on.cpp.o
  [ 11%] Building CXX object third_party/zstd/CMakeFiles/zstd.dir/decompress/zstd_decompress.cpp.o
  [ 12%] Building CXX object third_party/re2/CMakeFiles/re2.dir/parse.cpp.o
  [ 12%] Building CXX object third_party/zstd/CMakeFiles/zstd.dir/decompress/zstd_decompress_block.cpp.o
  [ 13%] Linking CXX static library libzstd.a
  [ 13%] Built target zstd
  [ 13%] Built target kuzu_extension
  [ 13%] Building CXX object src/binder/bind/copy/CMakeFiles/kuzu_binder_bind_copy.dir/bind_copy_rdf_graph.cpp.o
  [ 13%] Building CXX object src/binder/CMakeFiles/kuzu_binder.dir/binder_scope.cpp.o
  [ 13%] Building CXX object third_party/antlr4_runtime/CMakeFiles/antlr4_runtime.dir/src/atn/ATNConfigSet.cpp.o
  [ 13%] Building CXX object third_party/re2/CMakeFiles/re2.dir/perl_groups.cpp.o
  [ 13%] Building CXX object third_party/re2/CMakeFiles/re2.dir/prefilter.cpp.o
  [ 13%] Building CXX object third_party/re2/CMakeFiles/re2.dir/prefilter_tree.cpp.o
  [ 13%] Building CXX object src/binder/bind/ddl/CMakeFiles/kuzu_binder_bind_ddl.dir/bind_create_rdf_graph.cpp.o
  [ 13%] Building CXX object third_party/re2/CMakeFiles/re2.dir/prog.cpp.o
  [ 13%] Building CXX object src/binder/bind/copy/CMakeFiles/kuzu_binder_bind_copy.dir/bind_copy_to.cpp.o
  [ 14%] Building CXX object src/binder/CMakeFiles/kuzu_binder.dir/bound_statement_result.cpp.o
  [ 14%] Building CXX object third_party/antlr4_runtime/CMakeFiles/antlr4_runtime.dir/src/atn/ATN.cpp.o
  [ 14%] Building CXX object third_party/re2/CMakeFiles/re2.dir/re2.cpp.o
  [ 14%] Building CXX object src/binder/CMakeFiles/kuzu_binder.dir/bound_statement_rewriter.cpp.o
  [ 14%] Building CXX object third_party/antlr4_runtime/CMakeFiles/antlr4_runtime.dir/src/atn/ATNDeserializationOptions.cpp.o
  [ 14%] Building CXX object src/binder/bind/CMakeFiles/kuzu_binder_bind.dir/bind_create_macro.cpp.o
  [ 14%] Building CXX object third_party/re2/CMakeFiles/re2.dir/regexp.cpp.o
  [ 14%] Building CXX object third_party/antlr4_runtime/CMakeFiles/antlr4_runtime.dir/src/atn/ATNDeserializer.cpp.o
  [ 14%] Building CXX object third_party/re2/CMakeFiles/re2.dir/set.cpp.o
  [ 14%] Building CXX object src/binder/CMakeFiles/kuzu_binder.dir/bound_statement_visitor.cpp.o
  [ 14%] Built target kuzu_binder_bind_ddl
  [ 14%] Building CXX object third_party/antlr4_runtime/CMakeFiles/antlr4_runtime.dir/src/atn/ATNSimulator.cpp.o
  [ 14%] Linking CXX static library libfastpfor.a
  [ 15%] Building CXX object src/binder/bind/copy/CMakeFiles/kuzu_binder_bind_copy.dir/bind_copy_from.cpp.o
  [ 15%] Built target fastpfor
  [ 16%] Building CXX object src/binder/bind/CMakeFiles/kuzu_binder_bind.dir/bind_ddl.cpp.o
  [ 16%] Building CXX object third_party/re2/CMakeFiles/re2.dir/simplify.cpp.o
  [ 17%] Building CXX object third_party/re2/CMakeFiles/re2.dir/stringpiece.cpp.o
  [ 17%] Building CXX object src/binder/bind/read/CMakeFiles/kuzu_binder_bind_read.dir/bind_in_query_call.cpp.o
  [ 17%] Building CXX object third_party/re2/CMakeFiles/re2.dir/tostring.cpp.o
  [ 17%] Building CXX object src/binder/bind/CMakeFiles/kuzu_binder_bind.dir/bind_detach_database.cpp.o
  [ 17%] Building CXX object third_party/antlr4_runtime/CMakeFiles/antlr4_runtime.dir/src/atn/ATNState.cpp.o
  [ 17%] Building CXX object third_party/re2/CMakeFiles/re2.dir/unicode_casefold.cpp.o
  [ 17%] Building CXX object third_party/re2/CMakeFiles/re2.dir/unicode_groups.cpp.o
  [ 17%] Building CXX object third_party/re2/CMakeFiles/re2.dir/rune.cpp.o
  [ 17%] Building CXX object third_party/re2/CMakeFiles/re2.dir/strutil.cpp.o
  [ 17%] Building CXX object src/binder/CMakeFiles/kuzu_binder.dir/expression_binder.cpp.o
  [ 17%] Linking CXX static library libre2.a
  [ 17%] Built target re2
  [ 17%] Building CXX object src/binder/bind/read/CMakeFiles/kuzu_binder_bind_read.dir/bind_load_from.cpp.o
  [ 18%] Building CXX object third_party/antlr4_runtime/CMakeFiles/antlr4_runtime.dir/src/atn/ATNStateType.cpp.o
  [ 18%] Building CXX object src/binder/bind/CMakeFiles/kuzu_binder_bind.dir/bind_explain.cpp.o
  [ 18%] Building CXX object third_party/antlr4_runtime/CMakeFiles/antlr4_runtime.dir/src/atn/AtomTransition.cpp.o
  [ 18%] Building CXX object third_party/antlr4_runtime/CMakeFiles/antlr4_runtime.dir/src/atn/ContextSensitivityInfo.cpp.o
  [ 18%] Built target kuzu_binder_bind_copy
  [ 18%] Building CXX object src/binder/bind/read/CMakeFiles/kuzu_binder_bind_read.dir/bind_match.cpp.o
  [ 18%] Building CXX object src/binder/CMakeFiles/kuzu_binder.dir/expression_visitor.cpp.o
  [ 18%] Building CXX object third_party/antlr4_runtime/CMakeFiles/antlr4_runtime.dir/src/atn/DecisionEventInfo.cpp.o
  [ 18%] Building CXX object src/binder/bind/CMakeFiles/kuzu_binder_bind.dir/bind_file_scan.cpp.o
  [ 18%] Building CXX object src/binder/bind_expression/CMakeFiles/kuzu_binder_bind_expression.dir/bind_boolean_expression.cpp.o
  [ 18%] Building CXX object third_party/antlr4_runtime/CMakeFiles/antlr4_runtime.dir/src/atn/DecisionInfo.cpp.o
  [ 18%] Building CXX object src/binder/bind/CMakeFiles/kuzu_binder_bind.dir/bind_graph_pattern.cpp.o
  [ 18%] Building CXX object src/binder/bind/read/CMakeFiles/kuzu_binder_bind_read.dir/bind_unwind.cpp.o
  [ 18%] Built target kuzu_binder
  [ 18%] Building CXX object src/binder/bind_expression/CMakeFiles/kuzu_binder_bind_expression.dir/bind_case_expression.cpp.o
  [ 18%] Building CXX object third_party/antlr4_runtime/CMakeFiles/antlr4_runtime.dir/src/atn/DecisionState.cpp.o
  [ 18%] Building CXX object third_party/antlr4_runtime/CMakeFiles/antlr4_runtime.dir/src/atn/EpsilonTransition.cpp.o
  [ 18%] Building CXX object src/binder/bind_expression/CMakeFiles/kuzu_binder_bind_expression.dir/bind_comparison_expression.cpp.o
  [ 18%] Building CXX object src/binder/bind_expression/CMakeFiles/kuzu_binder_bind_expression.dir/bind_function_expression.cpp.o
  [ 18%] Building CXX object src/binder/bind_expression/CMakeFiles/kuzu_binder_bind_expression.dir/bind_literal_expression.cpp.o
  [ 18%] Building CXX object third_party/antlr4_runtime/CMakeFiles/antlr4_runtime.dir/src/atn/ErrorInfo.cpp.o
  [ 19%] Building CXX object third_party/antlr4_runtime/CMakeFiles/antlr4_runtime.dir/src/atn/LexerAction.cpp.o
  [ 19%] Building CXX object src/binder/expression/CMakeFiles/kuzu_binder_expression.dir/case_expression.cpp.o
  [ 19%] Building CXX object third_party/antlr4_runtime/CMakeFiles/antlr4_runtime.dir/src/atn/LexerActionExecutor.cpp.o
  [ 19%] Built target kuzu_binder_bind_read
  [ 19%] Building CXX object third_party/antlr4_runtime/CMakeFiles/antlr4_runtime.dir/src/atn/LexerATNConfig.cpp.o
  [ 20%] Building CXX object src/binder/expression/CMakeFiles/kuzu_binder_expression.dir/expression.cpp.o
  [ 20%] Building CXX object src/binder/expression/CMakeFiles/kuzu_binder_expression.dir/expression_util.cpp.o
  [ 20%] Building CXX object src/binder/bind/CMakeFiles/kuzu_binder_bind.dir/bind_projection_clause.cpp.o
  [ 20%] Building CXX object third_party/antlr4_runtime/CMakeFiles/antlr4_runtime.dir/src/atn/LexerATNSimulator.cpp.o
  [ 20%] Building CXX object src/binder/bind/CMakeFiles/kuzu_binder_bind.dir/bind_query.cpp.o
  [ 20%] Building CXX object src/binder/expression/CMakeFiles/kuzu_binder_expression.dir/function_expression.cpp.o
  [ 20%] Building CXX object src/binder/query/CMakeFiles/kuzu_binder_query.dir/bound_insert_clause.cpp.o
  [ 20%] Building CXX object src/binder/bind/CMakeFiles/kuzu_binder_bind.dir/bind_reading_clause.cpp.o
  [ 20%] Building CXX object third_party/antlr4_runtime/CMakeFiles/antlr4_runtime.dir/src/atn/LexerChannelAction.cpp.o
  [ 20%] Building CXX object src/binder/query/CMakeFiles/kuzu_binder_query.dir/bound_delete_clause.cpp.o
  [ 20%] Building CXX object src/binder/query/CMakeFiles/kuzu_binder_query.dir/bound_merge_clause.cpp.o
  [ 20%] Building CXX object src/binder/expression/CMakeFiles/kuzu_binder_expression.dir/literal_expression.cpp.o
  [ 20%] Building CXX object src/binder/rewriter/CMakeFiles/kuzu_binder_rewriter.dir/match_clause_pattern_label_rewriter.cpp.o
  [ 20%] Building CXX object third_party/antlr4_runtime/CMakeFiles/antlr4_runtime.dir/src/atn/LexerCustomAction.cpp.o
  [ 21%] Building CXX object src/binder/query/CMakeFiles/kuzu_binder_query.dir/bound_set_clause.cpp.o
  [ 21%] Building CXX object src/binder/bind_expression/CMakeFiles/kuzu_binder_bind_expression.dir/bind_null_operator_expression.cpp.o
  [ 21%] Building CXX object src/binder/expression/CMakeFiles/kuzu_binder_expression.dir/parameter_expression.cpp.o
  [ 21%] Building CXX object third_party/antlr4_runtime/CMakeFiles/antlr4_runtime.dir/src/atn/LexerIndexedCustomAction.cpp.o
  [ 21%] Building CXX object src/binder/query/CMakeFiles/kuzu_binder_query.dir/query_graph.cpp.o
  [ 21%] Building CXX object src/binder/visitor/CMakeFiles/kuzu_binder_visitor.dir/default_type_solver.cpp.o
  [ 21%] Building CXX object src/binder/expression/CMakeFiles/kuzu_binder_expression.dir/variable_expression.cpp.o
  [ 21%] Building CXX object src/binder/bind_expression/CMakeFiles/kuzu_binder_bind_expression.dir/bind_parameter_expression.cpp.o
  [ 21%] Building CXX object src/binder/rewriter/CMakeFiles/kuzu_binder_rewriter.dir/with_clause_projection_rewriter.cpp.o
  [ 21%] Building CXX object src/binder/bind/CMakeFiles/kuzu_binder_bind.dir/bind_standalone_call.cpp.o
  [ 21%] Building CXX object src/binder/visitor/CMakeFiles/kuzu_binder_visitor.dir/property_collector.cpp.o
  [ 21%] Built target kuzu_binder_expression
  [ 21%] Building CXX object third_party/antlr4_runtime/CMakeFiles/antlr4_runtime.dir/src/atn/LexerModeAction.cpp.o
  [ 21%] Building CXX object src/binder/query/CMakeFiles/kuzu_binder_query.dir/query_graph_label_analyzer.cpp.o
  [ 22%] Building CXX object src/binder/bind_expression/CMakeFiles/kuzu_binder_bind_expression.dir/bind_property_expression.cpp.o
  [ 23%] Building CXX object src/binder/bind/CMakeFiles/kuzu_binder_bind.dir/bind_transaction.cpp.o
  [ 23%] Building CXX object src/c_api/CMakeFiles/kuzu_c_api.dir/connection.cpp.o
  [ 23%] Built target kuzu_binder_rewriter
  [ 23%] Building CXX object third_party/antlr4_runtime/CMakeFiles/antlr4_runtime.dir/src/atn/LexerMoreAction.cpp.o
  [ 24%] Building CXX object src/catalog/CMakeFiles/kuzu_catalog.dir/catalog.cpp.o
  [ 25%] Building CXX object third_party/antlr4_runtime/CMakeFiles/antlr4_runtime.dir/src/atn/LexerPopModeAction.cpp.o
  [ 25%] Building CXX object third_party/antlr4_runtime/CMakeFiles/antlr4_runtime.dir/src/atn/LexerPushModeAction.cpp.o
  [ 26%] Building CXX object src/c_api/CMakeFiles/kuzu_c_api.dir/database.cpp.o
  [ 26%] Built target kuzu_binder_query
  [ 26%] Building CXX object src/binder/bind_expression/CMakeFiles/kuzu_binder_bind_expression.dir/bind_subquery_expression.cpp.o
  [ 26%] Built target kuzu_binder_visitor
  [ 26%] Building CXX object src/catalog/CMakeFiles/kuzu_catalog.dir/catalog_content.cpp.o
  [ 26%] Building CXX object src/catalog/CMakeFiles/kuzu_catalog.dir/catalog_set.cpp.o
  [ 26%] Building CXX object third_party/antlr4_runtime/CMakeFiles/antlr4_runtime.dir/src/atn/LexerSkipAction.cpp.o
  [ 26%] Building CXX object src/binder/bind/CMakeFiles/kuzu_binder_bind.dir/bind_updating_clause.cpp.o
  [ 26%] Building CXX object src/binder/bind/CMakeFiles/kuzu_binder_bind.dir/bind_extension.cpp.o
  [ 26%] Building CXX object third_party/antlr4_runtime/CMakeFiles/antlr4_runtime.dir/src/atn/LexerTypeAction.cpp.o
  [ 26%] Building CXX object src/c_api/CMakeFiles/kuzu_c_api.dir/data_type.cpp.o
  [ 26%] Building CXX object src/catalog/catalog_entry/CMakeFiles/kuzu_catalog_entry.dir/catalog_entry.cpp.o
  [ 26%] Building CXX object src/c_api/CMakeFiles/kuzu_c_api.dir/helpers.cpp.o
  [ 26%] Building CXX object third_party/antlr4_runtime/CMakeFiles/antlr4_runtime.dir/src/atn/LL1Analyzer.cpp.o
  [ 26%] Building CXX object src/c_api/CMakeFiles/kuzu_c_api.dir/flat_tuple.cpp.o
  [ 26%] Building CXX object src/common/CMakeFiles/kuzu_common.dir/case_insensitive_map.cpp.o
  [ 26%] Building CXX object src/c_api/CMakeFiles/kuzu_c_api.dir/prepared_statement.cpp.o
  [ 26%] Building CXX object third_party/antlr4_runtime/CMakeFiles/antlr4_runtime.dir/src/atn/LookaheadEventInfo.cpp.o
  [ 26%] Building CXX object src/binder/bind_expression/CMakeFiles/kuzu_binder_bind_expression.dir/bind_variable_expression.cpp.o
  [ 26%] Building CXX object src/catalog/CMakeFiles/kuzu_catalog.dir/property.cpp.o
  [ 26%] Building CXX object src/catalog/catalog_entry/CMakeFiles/kuzu_catalog_entry.dir/catalog_entry_type.cpp.o
  [ 26%] Building CXX object src/c_api/CMakeFiles/kuzu_c_api.dir/query_result.cpp.o
  [ 26%] Building CXX object src/common/CMakeFiles/kuzu_common.dir/constants.cpp.o
  [ 26%] Building CXX object src/common/CMakeFiles/kuzu_common.dir/expression_type.cpp.o
  [ 26%] Building CXX object third_party/antlr4_runtime/CMakeFiles/antlr4_runtime.dir/src/atn/NotSetTransition.cpp.o
  [ 26%] Building CXX object src/common/arrow/CMakeFiles/kuzu_common_arrow.dir/arrow_array_scan.cpp.o
  [ 26%] Building CXX object src/binder/bind/CMakeFiles/kuzu_binder_bind.dir/bind_export_database.cpp.o
  [ 26%] Building CXX object src/catalog/catalog_entry/CMakeFiles/kuzu_catalog_entry.dir/function_catalog_entry.cpp.o
  [ 26%] Building CXX object src/c_api/CMakeFiles/kuzu_c_api.dir/query_summary.cpp.o
  [ 26%] Building CXX object src/c_api/CMakeFiles/kuzu_c_api.dir/value.cpp.o
  [ 26%] Building CXX object third_party/antlr4_runtime/CMakeFiles/antlr4_runtime.dir/src/atn/OrderedATNConfigSet.cpp.o
  [ 26%] Built target kuzu_catalog
  [ 26%] Building CXX object src/catalog/catalog_entry/CMakeFiles/kuzu_catalog_entry.dir/table_catalog_entry.cpp.o
  [ 26%] Building CXX object src/common/arrow/CMakeFiles/kuzu_common_arrow.dir/arrow_converter.cpp.o
  [ 26%] Building CXX object src/common/CMakeFiles/kuzu_common.dir/in_mem_overflow_buffer.cpp.o
  [ 27%] Building CXX object third_party/antlr4_runtime/CMakeFiles/antlr4_runtime.dir/src/atn/ParseInfo.cpp.o
  [ 28%] Building CXX object src/common/CMakeFiles/kuzu_common.dir/logging_level_utils.cpp.o
  [ 28%] Building CXX object src/c_api/CMakeFiles/kuzu_c_api.dir/version.cpp.o
  [ 28%] Building CXX object src/common/arrow/CMakeFiles/kuzu_common_arrow.dir/arrow_null_mask_tree.cpp.o
  [ 28%] Built target kuzu_binder_bind_expression
  [ 29%] Building CXX object src/catalog/catalog_entry/CMakeFiles/kuzu_catalog_entry.dir/node_table_catalog_entry.cpp.o
  [ 29%] Building CXX object src/binder/bind/CMakeFiles/kuzu_binder_bind.dir/bind_import_database.cpp.o
  [ 29%] Built target kuzu_c_api
  [ 29%] Building CXX object third_party/antlr4_runtime/CMakeFiles/antlr4_runtime.dir/src/atn/ParserATNSimulator.cpp.o
  [ 29%] Building CXX object src/common/CMakeFiles/kuzu_common.dir/md5.cpp.o
  [ 29%] Building CXX object src/common/arrow/CMakeFiles/kuzu_common_arrow.dir/arrow_row_batch.cpp.o
  [ 29%] Building CXX object third_party/antlr4_runtime/CMakeFiles/antlr4_runtime.dir/src/atn/PrecedencePredicateTransition.cpp.o
  [ 30%] Building CXX object src/common/copier_config/CMakeFiles/kuzu_common_copier_config.dir/csv_reader_config.cpp.o
  [ 30%] Building CXX object src/common/CMakeFiles/kuzu_common.dir/metric.cpp.o
  [ 30%] Building CXX object src/catalog/catalog_entry/CMakeFiles/kuzu_catalog_entry.dir/rel_table_catalog_entry.cpp.o
  [ 30%] Building CXX object src/common/arrow/CMakeFiles/kuzu_common_arrow.dir/arrow_type.cpp.o
  [ 30%] Building CXX object third_party/antlr4_runtime/CMakeFiles/antlr4_runtime.dir/src/atn/PredicateEvalInfo.cpp.o
  [ 30%] Building CXX object src/binder/bind/CMakeFiles/kuzu_binder_bind.dir/bind_use_database.cpp.o
  [ 30%] Building CXX object src/catalog/catalog_entry/CMakeFiles/kuzu_catalog_entry.dir/rel_group_catalog_entry.cpp.o
  [ 30%] Building CXX object src/catalog/catalog_entry/CMakeFiles/kuzu_catalog_entry.dir/rdf_graph_catalog_entry.cpp.o
  [ 30%] Building CXX object third_party/antlr4_runtime/CMakeFiles/antlr4_runtime.dir/src/atn/PredicateTransition.cpp.o
  [ 30%] Built target kuzu_common_arrow
  [ 30%] Building CXX object src/common/CMakeFiles/kuzu_common.dir/null_mask.cpp.o
  [ 30%] Building CXX object src/catalog/catalog_entry/CMakeFiles/kuzu_catalog_entry.dir/scalar_macro_catalog_entry.cpp.o
  [ 30%] Building CXX object third_party/antlr4_runtime/CMakeFiles/antlr4_runtime.dir/src/atn/PredictionContextCache.cpp.o
  [ 30%] Building CXX object third_party/antlr4_runtime/CMakeFiles/antlr4_runtime.dir/src/atn/PredictionContext.cpp.o
  [ 30%] Building CXX object src/common/CMakeFiles/kuzu_common.dir/profiler.cpp.o
  [ 30%] Building CXX object src/common/CMakeFiles/kuzu_common.dir/type_utils.cpp.o
  [ 30%] Built target kuzu_binder_bind
  [ 30%] Building CXX object third_party/antlr4_runtime/CMakeFiles/antlr4_runtime.dir/src/atn/PredictionContextMergeCache.cpp.o
  [ 30%] Building CXX object src/common/CMakeFiles/kuzu_common.dir/utils.cpp.o
  [ 30%] Building CXX object src/common/CMakeFiles/kuzu_common.dir/sha256.cpp.o
  [ 30%] Building CXX object third_party/antlr4_runtime/CMakeFiles/antlr4_runtime.dir/src/atn/PredictionMode.cpp.o
  [ 31%] Building CXX object src/common/CMakeFiles/kuzu_common.dir/string_utils.cpp.o
  [ 31%] Building CXX object src/common/CMakeFiles/kuzu_common.dir/system_message.cpp.o
  [ 32%] Building CXX object third_party/antlr4_runtime/CMakeFiles/antlr4_runtime.dir/src/atn/ProfilingATNSimulator.cpp.o
  [ 32%] Built target kuzu_catalog_entry
  [ 32%] Building CXX object third_party/antlr4_runtime/CMakeFiles/antlr4_runtime.dir/src/atn/RangeTransition.cpp.o
  [ 32%] Building CXX object src/common/CMakeFiles/kuzu_common.dir/random_engine.cpp.o
  [ 32%] Building CXX object third_party/antlr4_runtime/CMakeFiles/antlr4_runtime.dir/src/atn/RuleTransition.cpp.o
  [ 32%] Building CXX object src/common/CMakeFiles/kuzu_common.dir/windows_utils.cpp.o
  [ 32%] Building CXX object third_party/antlr4_runtime/CMakeFiles/antlr4_runtime.dir/src/atn/SemanticContext.cpp.o
  [ 32%] Building CXX object third_party/antlr4_runtime/CMakeFiles/antlr4_runtime.dir/src/atn/SetTransition.cpp.o
  [ 32%] Building CXX object third_party/antlr4_runtime/CMakeFiles/antlr4_runtime.dir/src/atn/SingletonPredictionContext.cpp.o
  [ 32%] Building CXX object third_party/antlr4_runtime/CMakeFiles/antlr4_runtime.dir/src/atn/StarLoopbackState.cpp.o
  [ 32%] Building CXX object third_party/antlr4_runtime/CMakeFiles/antlr4_runtime.dir/src/atn/Transition.cpp.o
  [ 33%] Building CXX object third_party/antlr4_runtime/CMakeFiles/antlr4_runtime.dir/src/atn/TransitionType.cpp.o
  [ 33%] Building CXX object third_party/antlr4_runtime/CMakeFiles/antlr4_runtime.dir/src/atn/WildcardTransition.cpp.o
  [ 33%] Building CXX object third_party/antlr4_runtime/CMakeFiles/antlr4_runtime.dir/src/BailErrorStrategy.cpp.o
  [ 33%] Building CXX object third_party/antlr4_runtime/CMakeFiles/antlr4_runtime.dir/src/BaseErrorListener.cpp.o
  [ 33%] Building CXX object third_party/antlr4_runtime/CMakeFiles/antlr4_runtime.dir/src/BufferedTokenStream.cpp.o
  [ 33%] Building CXX object third_party/antlr4_runtime/CMakeFiles/antlr4_runtime.dir/src/CharStream.cpp.o
  [ 33%] Building CXX object third_party/antlr4_runtime/CMakeFiles/antlr4_runtime.dir/src/CommonToken.cpp.o
  [ 33%] Building CXX object third_party/antlr4_runtime/CMakeFiles/antlr4_runtime.dir/src/CommonTokenFactory.cpp.o
  [ 33%] Building CXX object third_party/antlr4_runtime/CMakeFiles/antlr4_runtime.dir/src/CommonTokenStream.cpp.o
  [ 34%] Building CXX object third_party/antlr4_runtime/CMakeFiles/antlr4_runtime.dir/src/ConsoleErrorListener.cpp.o
  [ 34%] Building CXX object third_party/antlr4_runtime/CMakeFiles/antlr4_runtime.dir/src/DefaultErrorStrategy.cpp.o
  [ 34%] Building CXX object third_party/antlr4_runtime/CMakeFiles/antlr4_runtime.dir/src/dfa/DFASerializer.cpp.o
  [ 34%] Building CXX object third_party/antlr4_runtime/CMakeFiles/antlr4_runtime.dir/src/dfa/DFA.cpp.o
  [ 34%] Building CXX object third_party/antlr4_runtime/CMakeFiles/antlr4_runtime.dir/src/dfa/DFAState.cpp.o
  [ 34%] Building CXX object third_party/antlr4_runtime/CMakeFiles/antlr4_runtime.dir/src/dfa/LexerDFASerializer.cpp.o
  [ 34%] Building CXX object third_party/antlr4_runtime/CMakeFiles/antlr4_runtime.dir/src/DiagnosticErrorListener.cpp.o
  [ 34%] Building CXX object third_party/antlr4_runtime/CMakeFiles/antlr4_runtime.dir/src/Exceptions.cpp.o
  [ 35%] Building CXX object third_party/antlr4_runtime/CMakeFiles/antlr4_runtime.dir/src/FailedPredicateException.cpp.o
  [ 35%] Building CXX object third_party/antlr4_runtime/CMakeFiles/antlr4_runtime.dir/src/InputMismatchException.cpp.o
  [ 35%] Building CXX object third_party/antlr4_runtime/CMakeFiles/antlr4_runtime.dir/src/internal/Synchronization.cpp.o
  [ 35%] Building CXX object third_party/antlr4_runtime/CMakeFiles/antlr4_runtime.dir/src/InterpreterRuleContext.cpp.o
  [ 35%] Building CXX object third_party/antlr4_runtime/CMakeFiles/antlr4_runtime.dir/src/IntStream.cpp.o
  [ 35%] Building CXX object third_party/antlr4_runtime/CMakeFiles/antlr4_runtime.dir/src/Lexer.cpp.o
  [ 35%] Building CXX object third_party/antlr4_runtime/CMakeFiles/antlr4_runtime.dir/src/LexerInterpreter.cpp.o
  [ 35%] Building CXX object third_party/antlr4_runtime/CMakeFiles/antlr4_runtime.dir/src/LexerNoViableAltException.cpp.o
  [ 35%] Building CXX object third_party/antlr4_runtime/CMakeFiles/antlr4_runtime.dir/src/ListTokenSource.cpp.o
  [ 36%] Building CXX object third_party/antlr4_runtime/CMakeFiles/antlr4_runtime.dir/src/misc/InterpreterDataReader.cpp.o
  [ 36%] Building CXX object third_party/antlr4_runtime/CMakeFiles/antlr4_runtime.dir/src/misc/Interval.cpp.o
  [ 36%] Building CXX object third_party/antlr4_runtime/CMakeFiles/antlr4_runtime.dir/src/misc/IntervalSet.cpp.o
  [ 36%] Building CXX object third_party/antlr4_runtime/CMakeFiles/antlr4_runtime.dir/src/misc/MurmurHash.cpp.o
  [ 36%] Building CXX object third_party/antlr4_runtime/CMakeFiles/antlr4_runtime.dir/src/misc/Predicate.cpp.o
  [ 36%] Building CXX object third_party/antlr4_runtime/CMakeFiles/antlr4_runtime.dir/src/NoViableAltException.cpp.o
  [ 36%] Building CXX object third_party/antlr4_runtime/CMakeFiles/antlr4_runtime.dir/src/Parser.cpp.o
  [ 36%] Building CXX object third_party/antlr4_runtime/CMakeFiles/antlr4_runtime.dir/src/ParserInterpreter.cpp.o
  [ 37%] Building CXX object third_party/antlr4_runtime/CMakeFiles/antlr4_runtime.dir/src/ParserRuleContext.cpp.o
  [ 37%] Building CXX object third_party/antlr4_runtime/CMakeFiles/antlr4_runtime.dir/src/ProxyErrorListener.cpp.o
  [ 37%] Building CXX object third_party/antlr4_runtime/CMakeFiles/antlr4_runtime.dir/src/RecognitionException.cpp.o
  [ 37%] Building CXX object third_party/antlr4_runtime/CMakeFiles/antlr4_runtime.dir/src/Recognizer.cpp.o
  [ 37%] Built target kuzu_common
  [ 37%] Building CXX object third_party/antlr4_runtime/CMakeFiles/antlr4_runtime.dir/src/RuleContext.cpp.o
  [ 37%] Building CXX object third_party/antlr4_runtime/CMakeFiles/antlr4_runtime.dir/src/RuleContextWithAltNum.cpp.o
  [ 37%] Building CXX object third_party/antlr4_runtime/CMakeFiles/antlr4_runtime.dir/src/RuntimeMetaData.cpp.o
  [ 37%] Building CXX object third_party/antlr4_runtime/CMakeFiles/antlr4_runtime.dir/src/support/Any.cpp.o
  [ 37%] Building CXX object third_party/antlr4_runtime/CMakeFiles/antlr4_runtime.dir/src/support/Arrays.cpp.o
  [ 38%] Building CXX object third_party/antlr4_runtime/CMakeFiles/antlr4_runtime.dir/src/support/CPPUtils.cpp.o
  [ 38%] Building CXX object third_party/antlr4_runtime/CMakeFiles/antlr4_runtime.dir/src/support/StringUtils.cpp.o
  [ 38%] Building CXX object third_party/antlr4_runtime/CMakeFiles/antlr4_runtime.dir/src/support/Utf8.cpp.o
  [ 38%] Building CXX object third_party/antlr4_runtime/CMakeFiles/antlr4_runtime.dir/src/Token.cpp.o
  [ 38%] Building CXX object third_party/antlr4_runtime/CMakeFiles/antlr4_runtime.dir/src/TokenSource.cpp.o
  [ 38%] Building CXX object third_party/antlr4_runtime/CMakeFiles/antlr4_runtime.dir/src/TokenStream.cpp.o
  [ 38%] Building CXX object third_party/antlr4_runtime/CMakeFiles/antlr4_runtime.dir/src/TokenStreamRewriter.cpp.o
  [ 38%] Building CXX object third_party/antlr4_runtime/CMakeFiles/antlr4_runtime.dir/src/tree/ErrorNodeImpl.cpp.o
  [ 39%] Building CXX object third_party/antlr4_runtime/CMakeFiles/antlr4_runtime.dir/src/tree/IterativeParseTreeWalker.cpp.o
  [ 39%] Building CXX object third_party/antlr4_runtime/CMakeFiles/antlr4_runtime.dir/src/tree/ParseTree.cpp.o
  [ 39%] Building CXX object third_party/antlr4_runtime/CMakeFiles/antlr4_runtime.dir/src/tree/ParseTreeListener.cpp.o
  [ 39%] Building CXX object third_party/antlr4_runtime/CMakeFiles/antlr4_runtime.dir/src/tree/ParseTreeVisitor.cpp.o
  [ 39%] Building CXX object third_party/antlr4_runtime/CMakeFiles/antlr4_runtime.dir/src/tree/ParseTreeWalker.cpp.o
  [ 39%] Building CXX object third_party/antlr4_runtime/CMakeFiles/antlr4_runtime.dir/src/tree/pattern/Chunk.cpp.o
  [ 39%] Building CXX object third_party/antlr4_runtime/CMakeFiles/antlr4_runtime.dir/src/tree/pattern/ParseTreeMatch.cpp.o
  [ 39%] Building CXX object third_party/antlr4_runtime/CMakeFiles/antlr4_runtime.dir/src/tree/pattern/ParseTreePattern.cpp.o
  [ 39%] Building CXX object third_party/antlr4_runtime/CMakeFiles/antlr4_runtime.dir/src/tree/pattern/ParseTreePatternMatcher.cpp.o
  [ 40%] Building CXX object third_party/antlr4_runtime/CMakeFiles/antlr4_runtime.dir/src/tree/pattern/RuleTagToken.cpp.o
  [ 40%] Building CXX object third_party/antlr4_runtime/CMakeFiles/antlr4_runtime.dir/src/tree/pattern/TagChunk.cpp.o
  [ 40%] Building CXX object third_party/antlr4_runtime/CMakeFiles/antlr4_runtime.dir/src/tree/pattern/TextChunk.cpp.o
  [ 40%] Building CXX object third_party/antlr4_runtime/CMakeFiles/antlr4_runtime.dir/src/tree/pattern/TokenTagToken.cpp.o
  [ 40%] Building CXX object third_party/antlr4_runtime/CMakeFiles/antlr4_runtime.dir/src/tree/TerminalNodeImpl.cpp.o
  [ 40%] Building CXX object third_party/antlr4_runtime/CMakeFiles/antlr4_runtime.dir/src/tree/Trees.cpp.o
  [ 40%] Building CXX object third_party/antlr4_runtime/CMakeFiles/antlr4_runtime.dir/src/tree/xpath/XPath.cpp.o
  [ 40%] Building CXX object third_party/antlr4_runtime/CMakeFiles/antlr4_runtime.dir/src/tree/xpath/XPathElement.cpp.o
  [ 41%] Building CXX object third_party/antlr4_runtime/CMakeFiles/antlr4_runtime.dir/src/tree/xpath/XPathLexer.cpp.o
  [ 41%] Building CXX object third_party/antlr4_runtime/CMakeFiles/antlr4_runtime.dir/src/tree/xpath/XPathLexerErrorListener.cpp.o
  [ 41%] Building CXX object third_party/antlr4_runtime/CMakeFiles/antlr4_runtime.dir/src/tree/xpath/XPathRuleAnywhereElement.cpp.o
  [ 41%] Building CXX object third_party/antlr4_runtime/CMakeFiles/antlr4_runtime.dir/src/tree/xpath/XPathRuleElement.cpp.o
  [ 41%] Building CXX object third_party/antlr4_runtime/CMakeFiles/antlr4_runtime.dir/src/tree/xpath/XPathTokenAnywhereElement.cpp.o
  [ 41%] Building CXX object third_party/antlr4_runtime/CMakeFiles/antlr4_runtime.dir/src/tree/xpath/XPathTokenElement.cpp.o
  [ 41%] Building CXX object third_party/antlr4_runtime/CMakeFiles/antlr4_runtime.dir/src/tree/xpath/XPathWildcardAnywhereElement.cpp.o
  [ 41%] Building CXX object third_party/antlr4_runtime/CMakeFiles/antlr4_runtime.dir/src/tree/xpath/XPathWildcardElement.cpp.o
  [ 41%] Building CXX object third_party/antlr4_runtime/CMakeFiles/antlr4_runtime.dir/src/UnbufferedCharStream.cpp.o
  [ 42%] Building CXX object third_party/antlr4_runtime/CMakeFiles/antlr4_runtime.dir/src/UnbufferedTokenStream.cpp.o
  [ 42%] Building CXX object third_party/antlr4_runtime/CMakeFiles/antlr4_runtime.dir/src/Vocabulary.cpp.o
  [ 42%] Building CXX object third_party/antlr4_runtime/CMakeFiles/antlr4_runtime.dir/src/WritableToken.cpp.o
  [ 42%] Linking CXX static library libantlr4_runtime.a
  [ 42%] Built target antlr4_runtime

  --- stderr
  -- Compiler is not EMSCRIPTEN
  CMake Deprecation Warning at third_party/re2/CMakeLists.txt:5 (cmake_minimum_required):
    Compatibility with CMake < 3.5 will be removed from a future version of
    CMake.

    Update the VERSION argument <min> value or use a ...<max> suffix to tell
    CMake that the project does not need compatibility with older versions.


  CMake Deprecation Warning at third_party/serd/CMakeLists.txt:4 (cmake_minimum_required):
    Compatibility with CMake < 3.5 will be removed from a future version of
    CMake.

    Update the VERSION argument <min> value or use a ...<max> suffix to tell
    CMake that the project does not need compatibility with older versions.


  CMake Warning:
    Manually-specified variables were not used by the project:

      CMAKE_ASM_COMPILER
      CMAKE_ASM_FLAGS


  In file included from /home/oleh/.cargo/registry/src/index.crates.io-6f17d22bba15001f/kuzu-0.4.2/kuzu-src/third_party/spdlog/spdlog/fmt/fmt.h:28,
                   from /home/oleh/.cargo/registry/src/index.crates.io-6f17d22bba15001f/kuzu-0.4.2/kuzu-src/third_party/spdlog/spdlog/common.h:50,
                   from /home/oleh/.cargo/registry/src/index.crates.io-6f17d22bba15001f/kuzu-0.4.2/kuzu-src/third_party/spdlog/spdlog/spdlog.h:12,
                   from /home/oleh/.cargo/registry/src/index.crates.io-6f17d22bba15001f/kuzu-0.4.2/kuzu-src/src/include/common/logging_level_utils.h:3,
                   from /home/oleh/.cargo/registry/src/index.crates.io-6f17d22bba15001f/kuzu-0.4.2/kuzu-src/src/common/logging_level_utils.cpp:1:
  /home/oleh/.cargo/registry/src/index.crates.io-6f17d22bba15001f/kuzu-0.4.2/kuzu-src/third_party/spdlog/spdlog/fmt/bundled/core.h: In instantiation of ‘constexpr fmt::v9::detail::value<Context> fmt::v9::detail::make_value(T&&) [with Context = fmt::v9::basic_format_context<fmt::v9::appender, char>; T = fmt::v9::basic_string_view<char>&]’:
  /home/oleh/.cargo/registry/src/index.crates.io-6f17d22bba15001f/kuzu-0.4.2/kuzu-src/third_party/spdlog/spdlog/fmt/bundled/core.h:1777:29:   required from ‘constexpr fmt::v9::format_arg_store<Context, Args>::format_arg_store(T&& ...) [with T = {fmt::v9::basic_string_view<char>&, const char (&)[3]}; Context = fmt::v9::basic_format_context<fmt::v9::appender, char>; Args = {fmt::v9::basic_string_view<char>, char [3]}]’
   1777 |   return make_value<Context>(val);
        |          ~~~~~~~~~~~~~~~~~~~^~~~~
  /home/oleh/.cargo/registry/src/index.crates.io-6f17d22bba15001f/kuzu-0.4.2/kuzu-src/third_party/spdlog/spdlog/fmt/bundled/core.h:1918:31:   required from ‘constexpr fmt::v9::format_arg_store<Context, typename std::remove_cv<typename std::remove_reference<Args>::type>::type ...> fmt::v9::make_format_args(Args&& ...) [with Context = basic_format_context<appender, char>; Args = {basic_string_view<char>&, const char (&)[3]}]’
   1918 |   return {FMT_FORWARD(args)...};
        |                               ^
  /home/oleh/.cargo/registry/src/index.crates.io-6f17d22bba15001f/kuzu-0.4.2/kuzu-src/third_party/spdlog/spdlog/fmt/bundled/core.h:3235:52:   required from ‘OutputIt fmt::v9::format_to(OutputIt, format_string<T ...>, T&& ...) [with OutputIt = appender; T = {basic_string_view<char>&, const char (&)[3]}; typename std::enable_if<detail::is_output_iterator<OutputIt, char>::value, int>::type <anonymous> = 0; format_string<T ...> = basic_format_string<char, basic_string_view<char>&, const char (&)[3]>]’
   3235 |   return vformat_to(out, fmt, fmt::make_format_args(args...));
        |                               ~~~~~~~~~~~~~~~~~~~~~^~~~~~~~~
  /home/oleh/.cargo/registry/src/index.crates.io-6f17d22bba15001f/kuzu-0.4.2/kuzu-src/third_party/spdlog/spdlog/fmt/bundled/format-inl.h:65:14:   required from here
     65 |     format_to(it, FMT_STRING("{}{}"), message, SEP);
        |     ~~~~~~~~~^~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
  /home/oleh/.cargo/registry/src/index.crates.io-6f17d22bba15001f/kuzu-0.4.2/kuzu-src/third_party/spdlog/spdlog/fmt/bundled/core.h:1901:77:   in ‘constexpr’ expansion of ‘fmt::v9::detail::make_arg<true, fmt::v9::basic_format_context<fmt::v9::appender, char>, fmt::v9::detail::type::string_type, fmt::v9::basic_string_view<char>&>((* & args#0))’
  /home/oleh/.cargo/registry/src/index.crates.io-6f17d22bba15001f/kuzu-0.4.2/kuzu-src/third_party/spdlog/spdlog/fmt/bundled/core.h:1735:15: warning: possibly dangling reference to a temporary [-Wdangling-reference]
   1735 |   const auto& arg = arg_mapper<Context>().map(FMT_FORWARD(val));
        |               ^~~
  /home/oleh/.cargo/registry/src/index.crates.io-6f17d22bba15001f/kuzu-0.4.2/kuzu-src/third_party/spdlog/spdlog/fmt/bundled/core.h:1735:46: note: the temporary was destroyed at the end of the full expression ‘fmt::v9::detail::arg_mapper<fmt::v9::basic_format_context<fmt::v9::appender, char> >().fmt::v9::detail::arg_mapper<fmt::v9::basic_format_context<fmt::v9::appender, char> >::map<fmt::v9::basic_string_view<char> >((*(const fmt::v9::basic_string_view<char>*)(& val)))’
   1735 |   const auto& arg = arg_mapper<Context>().map(FMT_FORWARD(val));
        |                     ~~~~~~~~~~~~~~~~~~~~~~~~~^~~~~~~~~~~~~~~~~~
  /home/oleh/.cargo/registry/src/index.crates.io-6f17d22bba15001f/kuzu-0.4.2/kuzu-src/src/common/copier_config/csv_reader_config.cpp: In instantiation of ‘bool kuzu::common::hasOption(const char* const (&)[size], const std::string&) [with long unsigned int size = 2; std::string = std::__cxx11::basic_string<char>]’:
  /home/oleh/.cargo/registry/src/index.crates.io-6f17d22bba15001f/kuzu-0.4.2/kuzu-src/src/common/copier_config/csv_reader_config.cpp:46:21:   required from here
     46 |     return hasOption(CopyConstants::BOOL_CSV_PARSING_OPTIONS, parsingOptionName);
        |            ~~~~~~~~~^~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
  /home/oleh/.cargo/registry/src/index.crates.io-6f17d22bba15001f/kuzu-0.4.2/kuzu-src/src/common/copier_config/csv_reader_config.cpp:42:21: error: no matching function for call to ‘find(const char* const*, const char* const*, const std::string&)’
     42 |     return std::find(std::begin(arr), std::end(arr), option) != std::end(arr);
        |            ~~~~~~~~~^~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
  In file included from /usr/include/c++/14/bits/locale_facets.h:48,
                   from /usr/include/c++/14/bits/basic_ios.h:37,
                   from /usr/include/c++/14/ios:46,
                   from /usr/include/c++/14/ostream:40,
                   from /usr/include/c++/14/bits/unique_ptr.h:43,
                   from /usr/include/c++/14/memory:78,
                   from /home/oleh/.cargo/registry/src/index.crates.io-6f17d22bba15001f/kuzu-0.4.2/kuzu-src/src/include/common/types/types.h:3,
                   from /home/oleh/.cargo/registry/src/index.crates.io-6f17d22bba15001f/kuzu-0.4.2/kuzu-src/src/include/common/types/ku_list.h:3,
                   from /home/oleh/.cargo/registry/src/index.crates.io-6f17d22bba15001f/kuzu-0.4.2/kuzu-src/src/include/common/types/value/value.h:10,
                   from /home/oleh/.cargo/registry/src/index.crates.io-6f17d22bba15001f/kuzu-0.4.2/kuzu-src/src/include/common/copier_config/csv_reader_config.h:5,
                   from /home/oleh/.cargo/registry/src/index.crates.io-6f17d22bba15001f/kuzu-0.4.2/kuzu-src/src/common/copier_config/csv_reader_config.cpp:1:
  /usr/include/c++/14/bits/streambuf_iterator.h:435:5: note: candidate: ‘template<class _CharT2> typename __gnu_cxx::__enable_if<std::__is_char<_CharT2>::__value, std::istreambuf_iterator<_CharT, std::char_traits<_CharT> > >::__type std::find(istreambuf_iterator<_CharT, char_traits<_CharT> >, istreambuf_iterator<_CharT, char_traits<_CharT> >, const _CharT2&)’
    435 |     find(istreambuf_iterator<_CharT> __first,
        |     ^~~~
  /usr/include/c++/14/bits/streambuf_iterator.h:435:5: note:   template argument deduction/substitution failed:
  /home/oleh/.cargo/registry/src/index.crates.io-6f17d22bba15001f/kuzu-0.4.2/kuzu-src/src/common/copier_config/csv_reader_config.cpp:42:21: note:   mismatched types ‘std::istreambuf_iterator<_CharT, std::char_traits<_CharT> >’ and ‘const char* const*’
     42 |     return std::find(std::begin(arr), std::end(arr), option) != std::end(arr);
        |            ~~~~~~~~~^~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
  /home/oleh/.cargo/registry/src/index.crates.io-6f17d22bba15001f/kuzu-0.4.2/kuzu-src/src/common/copier_config/csv_reader_config.cpp: In instantiation of ‘bool kuzu::common::hasOption(const char* const (&)[size], const std::string&) [with long unsigned int size = 3; std::string = std::__cxx11::basic_string<char>]’:
  /home/oleh/.cargo/registry/src/index.crates.io-6f17d22bba15001f/kuzu-0.4.2/kuzu-src/src/common/copier_config/csv_reader_config.cpp:50:21:   required from here
     50 |     return hasOption(CopyConstants::STRING_CSV_PARSING_OPTIONS, parsingOptionName);
        |            ~~~~~~~~~^~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
  /home/oleh/.cargo/registry/src/index.crates.io-6f17d22bba15001f/kuzu-0.4.2/kuzu-src/src/common/copier_config/csv_reader_config.cpp:42:21: error: no matching function for call to ‘find(const char* const*, const char* const*, const std::string&)’
     42 |     return std::find(std::begin(arr), std::end(arr), option) != std::end(arr);
        |            ~~~~~~~~~^~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
  /usr/include/c++/14/bits/streambuf_iterator.h:435:5: note: candidate: ‘template<class _CharT2> typename __gnu_cxx::__enable_if<std::__is_char<_CharT2>::__value, std::istreambuf_iterator<_CharT, std::char_traits<_CharT> > >::__type std::find(istreambuf_iterator<_CharT, char_traits<_CharT> >, istreambuf_iterator<_CharT, char_traits<_CharT> >, const _CharT2&)’
    435 |     find(istreambuf_iterator<_CharT> __first,
        |     ^~~~
  /usr/include/c++/14/bits/streambuf_iterator.h:435:5: note:   template argument deduction/substitution failed:
  /home/oleh/.cargo/registry/src/index.crates.io-6f17d22bba15001f/kuzu-0.4.2/kuzu-src/src/common/copier_config/csv_reader_config.cpp:42:21: note:   mismatched types ‘std::istreambuf_iterator<_CharT, std::char_traits<_CharT> >’ and ‘const char* const*’
     42 |     return std::find(std::begin(arr), std::end(arr), option) != std::end(arr);
        |            ~~~~~~~~~^~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
  gmake[2]: *** [src/common/copier_config/CMakeFiles/kuzu_common_copier_config.dir/build.make:76: src/common/copier_config/CMakeFiles/kuzu_common_copier_config.dir/csv_reader_config.cpp.o] Error 1
  gmake[1]: *** [CMakeFiles/Makefile2:3713: src/common/copier_config/CMakeFiles/kuzu_common_copier_config.dir/all] Error 2
  gmake[1]: *** Waiting for unfinished jobs....
  In file included from /home/oleh/.cargo/registry/src/index.crates.io-6f17d22bba15001f/kuzu-0.4.2/kuzu-src/third_party/spdlog/spdlog/fmt/fmt.h:28,
                   from /home/oleh/.cargo/registry/src/index.crates.io-6f17d22bba15001f/kuzu-0.4.2/kuzu-src/third_party/spdlog/spdlog/common.h:50,
                   from /home/oleh/.cargo/registry/src/index.crates.io-6f17d22bba15001f/kuzu-0.4.2/kuzu-src/third_party/spdlog/spdlog/details/registry.h:11,
                   from /home/oleh/.cargo/registry/src/index.crates.io-6f17d22bba15001f/kuzu-0.4.2/kuzu-src/third_party/spdlog/spdlog/details/synchronous_factory.h:6,
                   from /home/oleh/.cargo/registry/src/index.crates.io-6f17d22bba15001f/kuzu-0.4.2/kuzu-src/third_party/spdlog/spdlog/sinks/stdout_sinks.h:8,
                   from /home/oleh/.cargo/registry/src/index.crates.io-6f17d22bba15001f/kuzu-0.4.2/kuzu-src/src/common/utils.cpp:3:
  /home/oleh/.cargo/registry/src/index.crates.io-6f17d22bba15001f/kuzu-0.4.2/kuzu-src/third_party/spdlog/spdlog/fmt/bundled/core.h: In instantiation of ‘constexpr fmt::v9::detail::value<Context> fmt::v9::detail::make_value(T&&) [with Context = fmt::v9::basic_format_context<fmt::v9::appender, char>; T = fmt::v9::basic_string_view<char>&]’:
  /home/oleh/.cargo/registry/src/index.crates.io-6f17d22bba15001f/kuzu-0.4.2/kuzu-src/third_party/spdlog/spdlog/fmt/bundled/core.h:1777:29:   required from ‘constexpr fmt::v9::format_arg_store<Context, Args>::format_arg_store(T&& ...) [with T = {fmt::v9::basic_string_view<char>&, const char (&)[3]}; Context = fmt::v9::basic_format_context<fmt::v9::appender, char>; Args = {fmt::v9::basic_string_view<char>, char [3]}]’
   1777 |   return make_value<Context>(val);
        |          ~~~~~~~~~~~~~~~~~~~^~~~~
  /home/oleh/.cargo/registry/src/index.crates.io-6f17d22bba15001f/kuzu-0.4.2/kuzu-src/third_party/spdlog/spdlog/fmt/bundled/core.h:1918:31:   required from ‘constexpr fmt::v9::format_arg_store<Context, typename std::remove_cv<typename std::remove_reference<Args>::type>::type ...> fmt::v9::make_format_args(Args&& ...) [with Context = basic_format_context<appender, char>; Args = {basic_string_view<char>&, const char (&)[3]}]’
   1918 |   return {FMT_FORWARD(args)...};
        |                               ^
  /home/oleh/.cargo/registry/src/index.crates.io-6f17d22bba15001f/kuzu-0.4.2/kuzu-src/third_party/spdlog/spdlog/fmt/bundled/core.h:3235:52:   required from ‘OutputIt fmt::v9::format_to(OutputIt, format_string<T ...>, T&& ...) [with OutputIt = appender; T = {basic_string_view<char>&, const char (&)[3]}; typename std::enable_if<detail::is_output_iterator<OutputIt, char>::value, int>::type <anonymous> = 0; format_string<T ...> = basic_format_string<char, basic_string_view<char>&, const char (&)[3]>]’
   3235 |   return vformat_to(out, fmt, fmt::make_format_args(args...));
        |                               ~~~~~~~~~~~~~~~~~~~~~^~~~~~~~~
  /home/oleh/.cargo/registry/src/index.crates.io-6f17d22bba15001f/kuzu-0.4.2/kuzu-src/third_party/spdlog/spdlog/fmt/bundled/format-inl.h:65:14:   required from here
     65 |     format_to(it, FMT_STRING("{}{}"), message, SEP);
        |     ~~~~~~~~~^~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
  /home/oleh/.cargo/registry/src/index.crates.io-6f17d22bba15001f/kuzu-0.4.2/kuzu-src/third_party/spdlog/spdlog/fmt/bundled/core.h:1901:77:   in ‘constexpr’ expansion of ‘fmt::v9::detail::make_arg<true, fmt::v9::basic_format_context<fmt::v9::appender, char>, fmt::v9::detail::type::string_type, fmt::v9::basic_string_view<char>&>((* & args#0))’
  /home/oleh/.cargo/registry/src/index.crates.io-6f17d22bba15001f/kuzu-0.4.2/kuzu-src/third_party/spdlog/spdlog/fmt/bundled/core.h:1735:15: warning: possibly dangling reference to a temporary [-Wdangling-reference]
   1735 |   const auto& arg = arg_mapper<Context>().map(FMT_FORWARD(val));
        |               ^~~
  /home/oleh/.cargo/registry/src/index.crates.io-6f17d22bba15001f/kuzu-0.4.2/kuzu-src/third_party/spdlog/spdlog/fmt/bundled/core.h:1735:46: note: the temporary was destroyed at the end of the full expression ‘fmt::v9::detail::arg_mapper<fmt::v9::basic_format_context<fmt::v9::appender, char> >().fmt::v9::detail::arg_mapper<fmt::v9::basic_format_context<fmt::v9::appender, char> >::map<fmt::v9::basic_string_view<char> >((*(const fmt::v9::basic_string_view<char>*)(& val)))’
   1735 |   const auto& arg = arg_mapper<Context>().map(FMT_FORWARD(val));
        |                     ~~~~~~~~~~~~~~~~~~~~~~~~~^~~~~~~~~~~~~~~~~~
  /home/oleh/.cargo/registry/src/index.crates.io-6f17d22bba15001f/kuzu-0.4.2/kuzu-src/third_party/antlr4_runtime/src/Vocabulary.cpp: In static member function ‘static void antlr4::dfa::Vocabulary::__static_initialization_and_destruction_0()’:
  /home/oleh/.cargo/registry/src/index.crates.io-6f17d22bba15001f/kuzu-0.4.2/kuzu-src/third_party/antlr4_runtime/src/Vocabulary.cpp:12:18: warning: ‘antlr4::dfa::Vocabulary::EMPTY_VOCABULARY’ is deprecated: Use the default constructor of Vocabulary instead. [-Wdeprecated-declarations]
     12 | const Vocabulary Vocabulary::EMPTY_VOCABULARY;
        |                  ^~~~~~~~~~
  /home/oleh/.cargo/registry/src/index.crates.io-6f17d22bba15001f/kuzu-0.4.2/kuzu-src/third_party/antlr4_runtime/src/Vocabulary.cpp:12:18: note: declared here
  gmake: *** [Makefile:136: all] Error 2
  thread 'main' panicked at /home/oleh/.cargo/registry/src/index.crates.io-6f17d22bba15001f/cmake-0.1.50/src/lib.rs:1098:5:

  command did not execute successfully, got: exit status: 2

  build script failed, must exit now
  note: run with `RUST_BACKTRACE=1` environment variable to display a backtrace
```

--- a/src/include/storage/index/hash_index.h
+++ b/src/include/storage/index/hash_index.h
@@ -1,5 +1,6 @@
 #pragma once
 
+#include <algorithm>
 #include <cstdint>
 #include <string_view>
 #include <type_traits>
@@ -13,6 +14,7 @@
 #include "hash_index_slot.h"
 #include "storage/index/hash_index_utils.h"
 #include "storage/index/in_mem_hash_index.h"
+#include "storage/local_storage/local_hash_index.h"
 
 namespace kuzu {
 namespace common {
@@ -30,9 +32,6 @@ class OverflowFileHandle;
 template<typename T>
 class DiskArray;
 class DiskArrayCollection;
-
-template<typename T>
-class HashIndexLocalStorage;
 
 class OnDiskHashIndex {
 public:
@@ -78,15 +77,71 @@ public:
 public:
     using Key =
         typename std::conditional<std::same_as<T, common::ku_string_t>, std::string_view, T>::type;
-    bool lookupInternal(transaction::Transaction* transaction, Key key, common::offset_t& result);
-    void deleteInternal(Key key) const;
-    bool insertInternal(Key key, common::offset_t value);
+    // For read transactions, local storage is skipped, lookups are performed on the persistent
+    // storage. For write transactions, lookups are performed in the local storage first, then in
+    // the persistent storage if necessary. In details, there are three cases for the local storage
+    // lookup:
+    // - the key is found in the local storage, directly return true;
+    // - the key has been marked as deleted in the local storage, return false;
+    // - the key is neither deleted nor found in the local storage, lookup in the persistent
+    // storage.
+    bool lookupInternal(transaction::Transaction* transaction, Key key, common::offset_t& result) {
+        if (transaction->isReadOnly()) {
+            return lookupInPersistentIndex(transaction->getType(), key, result);
+        } else {
+            KU_ASSERT(transaction->isWriteTransaction());
+            auto localLookupState = localStorage->lookup(key, result);
+            if (localLookupState == HashIndexLocalLookupState::KEY_DELETED) {
+                return false;
+            } else if (localLookupState == HashIndexLocalLookupState::KEY_FOUND) {
+                return true;
+            } else {
+                KU_ASSERT(localLookupState == HashIndexLocalLookupState::KEY_NOT_EXIST);
+                return lookupInPersistentIndex(transaction->getType(), key, result);
+            }
+        }
+    }
+
+    // For deletions, we don't check if the deleted keys exist or not. Thus, we don't need to check
+    // in the persistent storage and directly delete keys in the local storage.
+    void deleteInternal(Key key) const { localStorage->deleteKey(key); }
+
+    // For insertions, we first check in the local storage. There are three cases:
+    // - the key is found in the local storage, return false;
+    // - the key is marked as deleted in the local storage, insert the key to the local storage;
+    // - the key doesn't exist in the local storage, check if the key exists in the persistent
+    // index, if
+    //   so, return false, else insert the key to the local storage.
+    bool insertInternal(Key key, common::offset_t value) {
+        common::offset_t tmpResult;
+        auto localLookupState = localStorage->lookup(key, tmpResult);
+        if (localLookupState == HashIndexLocalLookupState::KEY_FOUND) {
+            return false;
+        } else if (localLookupState == HashIndexLocalLookupState::KEY_NOT_EXIST) {
+            if (lookupInPersistentIndex(transaction::TransactionType::WRITE, key, tmpResult)) {
+                return false;
+            }
+        }
+        return localStorage->insert(key, value);
+    }
 
     using BufferKeyType =
         typename std::conditional<std::same_as<T, common::ku_string_t>, std::string, T>::type;
     // Appends the buffer to the index. Returns the number of values successfully inserted,
     // or the index of the first value which cannot be inserted.
-    size_t append(const IndexBuffer<BufferKeyType>& buffer);
+    size_t append(const IndexBuffer<BufferKeyType>& buffer) {
+        // Check if values already exist in persistent storage
+        if (indexHeaderForWriteTrx.numEntries > 0) {
+            common::offset_t result;
+            for (size_t i = 0; i < buffer.size(); i++) {
+                const auto& [key, value] = buffer[i];
+                if (lookupInPersistentIndex(transaction::TransactionType::WRITE, key, result)) {
+                    return i;
+                }
+            }
+        }
+        return localStorage->append(buffer);
+    }
 
     bool prepareCommit() override;
     void prepareRollback() override;
@@ -96,11 +151,40 @@ public:
 
 private:
     bool lookupInPersistentIndex(transaction::TransactionType trxType, Key key,
-        common::offset_t& result);
+        common::offset_t& result) {
+        auto& header = trxType == transaction::TransactionType::READ_ONLY ?
+                           this->indexHeaderForReadTrx :
+                           this->indexHeaderForWriteTrx;
+        // There may not be any primary key slots if we try to lookup on an empty index
+        if (header.numEntries == 0) {
+            return false;
+        }
+        auto hashValue = HashIndexUtils::hash(key);
+        auto fingerprint = HashIndexUtils::getFingerprintForHash(hashValue);
+        auto iter =
+            getSlotIterator(HashIndexUtils::getPrimarySlotIdForHash(header, hashValue), trxType);
+        do {
+            auto entryPos = findMatchedEntryInSlot(trxType, iter.slot, key, fingerprint);
+            if (entryPos != SlotHeader::INVALID_ENTRY_POS) {
+                result = iter.slot.entries[entryPos].value;
+                return true;
+            }
+        } while (nextChainedSlot(trxType, iter));
+        return false;
+    }
     void deleteFromPersistentIndex(Key key);
 
     entry_pos_t findMatchedEntryInSlot(transaction::TransactionType trxType, const Slot<T>& slot,
-        Key key, uint8_t fingerprint) const;
+        Key key, uint8_t fingerprint) const {
+        for (auto entryPos = 0u; entryPos < getSlotCapacity<T>(); entryPos++) {
+            if (slot.header.isEntryValid(entryPos) &&
+                slot.header.fingerprints[entryPos] == fingerprint &&
+                equals(trxType, key, slot.entries[entryPos].key)) {
+                return entryPos;
+            }
+        }
+        return SlotHeader::INVALID_ENTRY_POS;
+    }
 
     inline void updateSlot(const SlotInfo& slotInfo, const Slot<T>& slot) {
         slotInfo.slotType == SlotType::PRIMARY ? pSlots->update(slotInfo.slotId, slot) :
@@ -191,7 +275,13 @@ common::hash_t HashIndex<common::ku_string_t>::hashStored(transaction::Transacti
 
 template<>
 inline bool HashIndex<common::ku_string_t>::equals(transaction::TransactionType trxType,
-    std::string_view keyToLookup, const common::ku_string_t& keyInEntry) const;
+    std::string_view keyToLookup, const common::ku_string_t& keyInEntry) const {
+    if (HashIndexUtils::areStringPrefixAndLenEqual(keyToLookup, keyInEntry)) {
+        auto entryKeyString = overflowFileHandle->readString(trxType, keyInEntry);
+        return memcmp(keyToLookup.data(), entryKeyString.c_str(), entryKeyString.length()) == 0;
+    }
+    return false;
+}
 
 class PrimaryKeyIndex {
 public:

--- a/src/include/storage/index/in_mem_hash_index.h
+++ b/src/include/storage/index/in_mem_hash_index.h
@@ -81,12 +81,44 @@ public:
     using Key = std::conditional_t<std::same_as<T, common::ku_string_t>, std::string_view, T>;
     // Appends the buffer to the index. Returns the number of values successfully inserted.
     // I.e. if a key fails to insert, its index will be the return value
-    size_t append(const IndexBuffer<BufferKeyType>& buffer);
+    size_t append(const IndexBuffer<BufferKeyType>& buffer) {
+        reserve(indexHeader.numEntries + buffer.size());
+        // Do both searches after splitting. Returning early if the key already exists isn't a
+        // particular concern and doing both after splitting allows the slotID to be reused
+        common::hash_t hashes[BUFFER_SIZE];
+        for (size_t i = 0; i < buffer.size(); i++) {
+            hashes[i] = HashIndexUtils::hash(buffer[i].first);
+        }
+        for (size_t i = 0; i < buffer.size(); i++) {
+            auto& [key, value] = buffer[i];
+            if (!appendInternal(key, value, hashes[i])) {
+                return i;
+            }
+        }
+        return buffer.size();
+    }
+
     inline bool append(Key key, common::offset_t value) {
         reserve(indexHeader.numEntries + 1);
         return appendInternal(key, value, HashIndexUtils::hash(key));
     }
-    bool lookup(Key key, common::offset_t& result);
+    bool lookup(Key key, common::offset_t& result) {
+        // This needs to be fast if the builder is empty since this function is always tried
+        // when looking up in the persistent hash index
+        if (this->indexHeader.numEntries == 0) {
+            return false;
+        }
+        auto hashValue = HashIndexUtils::hash(key);
+        auto fingerprint = HashIndexUtils::getFingerprintForHash(hashValue);
+        auto slotId = HashIndexUtils::getPrimarySlotIdForHash(this->indexHeader, hashValue);
+        SlotIterator iter(slotId, this);
+        auto entryPos = findEntry(iter, key, fingerprint);
+        if (entryPos != SlotHeader::INVALID_ENTRY_POS) {
+            result = iter.slot->entries[entryPos].value;
+            return true;
+        }
+        return false;
+    }
 
     uint64_t size() const { return this->indexHeader.numEntries; }
     bool empty() const { return size() == 0; }
@@ -118,11 +150,71 @@ public:
 
     // Deletes key, maintaining gapless structure by replacing it with the last entry in the
     // slot
-    bool deleteKey(Key key);
+    bool deleteKey(Key key) {
+        if (this->indexHeader.numEntries == 0) {
+            return false;
+        }
+        auto hashValue = HashIndexUtils::hash(key);
+        auto fingerprint = HashIndexUtils::getFingerprintForHash(hashValue);
+        auto slotId = HashIndexUtils::getPrimarySlotIdForHash(this->indexHeader, hashValue);
+        SlotIterator iter(slotId, this);
+        std::optional<entry_pos_t> deletedPos = 0;
+        do {
+            for (auto entryPos = 0u; entryPos < getSlotCapacity<T>(); entryPos++) {
+                if (iter.slot->header.isEntryValid(entryPos) &&
+                    iter.slot->header.fingerprints[entryPos] == fingerprint &&
+                    equals(key, iter.slot->entries[entryPos].key)) {
+                    deletedPos = entryPos;
+                    iter.slot->header.setEntryInvalid(entryPos);
+                    break;
+                }
+            }
+            if (deletedPos.has_value()) {
+                break;
+            }
+        } while (nextChainedSlot(iter));
+
+        if (deletedPos.has_value()) {
+            // Find the last valid entry and move it into the deleted position
+            auto newIter = iter;
+            while (nextChainedSlot(newIter))
+                ;
+            if (newIter.slotInfo != iter.slotInfo ||
+                *deletedPos != newIter.slot->header.numEntries() - 1) {
+                auto lastEntryPos = newIter.slot->header.numEntries();
+                iter.slot->entries[*deletedPos] = newIter.slot->entries[lastEntryPos];
+                iter.slot->header.setEntryValid(*deletedPos,
+                    newIter.slot->header.fingerprints[lastEntryPos]);
+                newIter.slot->header.setEntryInvalid(lastEntryPos);
+            }
+        }
+        return false;
+    }
 
 private:
     // Assumes that space has already been allocated for the entry
-    bool appendInternal(Key key, common::offset_t value, common::hash_t hash);
+    bool appendInternal(Key key, common::offset_t value, common::hash_t hash) {
+        auto fingerprint = HashIndexUtils::getFingerprintForHash(hash);
+        auto slotID = HashIndexUtils::getPrimarySlotIdForHash(this->indexHeader, hash);
+        SlotIterator iter(slotID, this);
+        // The builder never keeps holes and doesn't support deletions
+        // Check the valid entries, then insert at the end if we don't find one which matches
+        auto entryPos = findEntry(iter, key, fingerprint);
+        auto numEntries = iter.slot->header.numEntries();
+        if (entryPos != SlotHeader::INVALID_ENTRY_POS) {
+            // The key already exists
+            return false;
+        } else if (numEntries < getSlotCapacity<T>()) [[likely]] {
+            // The key does not exist and the last slot has free space
+            insert(key, iter.slot, numEntries, value, fingerprint);
+            this->indexHeader.numEntries++;
+            return true;
+        }
+        // The last slot is full. Insert a new one
+        insertToNewOvfSlot(key, iter.slot, value, fingerprint);
+        this->indexHeader.numEntries++;
+        return true;
+    }
     Slot<T>* getSlot(const SlotInfo& slotInfo) const;
 
     uint32_t allocatePSlots(uint32_t numSlotsToAllocate);
@@ -138,24 +230,73 @@ private:
     void reclaimOverflowSlots(SlotIterator iter);
 
     inline bool equals(Key keyToLookup, const T& keyInEntry) const {
-        return keyToLookup == keyInEntry;
+        if constexpr (std::same_as<T, common::ku_string_t>) {
+            // Checks if prefix and len matches first.
+            if (!HashIndexUtils::areStringPrefixAndLenEqual(keyToLookup, keyInEntry)) {
+                return false;
+            }
+            if (keyInEntry.len <= common::ku_string_t::PREFIX_LENGTH) {
+                // For strings shorter than PREFIX_LENGTH, the result must be true.
+                return true;
+            } else if (keyInEntry.len <= common::ku_string_t::SHORT_STR_LENGTH) {
+                // For short strings, whose lengths are larger than PREFIX_LENGTH, check if their
+                // actual values are equal.
+                return memcmp(keyToLookup.data(), keyInEntry.prefix, keyInEntry.len) == 0;
+            } else {
+                // For long strings, compare with overflow data
+                return overflowFileHandle->equals(transaction::TransactionType::WRITE, keyToLookup,
+                    keyInEntry);
+            }
+        } else {
+            return keyToLookup == keyInEntry;
+        }
     }
 
     inline void insert(Key key, Slot<T>* slot, uint8_t entryPos, common::offset_t value,
         uint8_t fingerprint) {
-        slot->entries[entryPos] = SlotEntry<T>(key, value);
-        slot->header.setEntryValid(entryPos, fingerprint);
         KU_ASSERT(HashIndexUtils::getFingerprintForHash(HashIndexUtils::hash(key)) == fingerprint);
+        auto& entry = slot->entries[entryPos];
+        if constexpr (std::same_as<T, common::ku_string_t>) {
+            entry = SlotEntry<common::ku_string_t>(overflowFileHandle->writeString(key), value);
+            slot->header.setEntryValid(entryPos, fingerprint);
+        } else {
+            entry = SlotEntry<T>(key, value);
+            slot->header.setEntryValid(entryPos, fingerprint);
+        }
     }
     void copy(const SlotEntry<T>& oldEntry, slot_id_t newSlotId, uint8_t fingerprint);
+
     void insertToNewOvfSlot(Key key, Slot<T>* previousSlot, common::offset_t offset,
-        uint8_t fingerprint);
+        uint8_t fingerprint) {
+        auto newSlotId = allocateAOSlot();
+        previousSlot->header.nextOvfSlotId = newSlotId;
+        auto newSlot = getSlot(SlotInfo{newSlotId, SlotType::OVF});
+        auto entryPos = 0u; // Always insert to the first entry when there is a new slot.
+        insert(key, newSlot, entryPos, offset, fingerprint);
+    }
+
     common::hash_t hashStored(const T& key) const;
     Slot<T>* clearNextOverflowAndAdvanceIter(SlotIterator& iter);
 
     // Finds the entry matching the given key. The iterator will be advanced and will either point
     // to the slot containing the matching entry, or the last slot available
-    entry_pos_t findEntry(SlotIterator& iter, Key key, uint8_t fingerprint);
+    entry_pos_t findEntry(SlotIterator& iter, Key key, uint8_t fingerprint) {
+        do {
+            auto numEntries = iter.slot->header.numEntries();
+            KU_ASSERT(numEntries == std::countr_one(iter.slot->header.validityMask));
+            for (auto entryPos = 0u; entryPos < numEntries; entryPos++) {
+                if (iter.slot->header.fingerprints[entryPos] == fingerprint &&
+                    equals(key, iter.slot->entries[entryPos].key)) [[unlikely]] {
+                    // Value already exists
+                    return entryPos;
+                }
+            }
+            if (numEntries < getSlotCapacity<T>()) {
+                return SlotHeader::INVALID_ENTRY_POS;
+            }
+        } while (nextChainedSlot(iter));
+        return SlotHeader::INVALID_ENTRY_POS;
+    }
 
 private:
     // TODO: might be more efficient to use a vector for each slot since this is now only needed
@@ -166,8 +307,5 @@ private:
     HashIndexHeader indexHeader;
 };
 
-template<>
-bool InMemHashIndex<common::ku_string_t>::equals(std::string_view keyToLookup,
-    const common::ku_string_t& keyInEntry) const;
 } // namespace storage
 } // namespace kuzu

--- a/src/include/storage/store/column.h
+++ b/src/include/storage/store/column.h
@@ -1,5 +1,7 @@
 #pragma once
 
+#include <algorithm>
+
 #include "catalog/catalog.h"
 #include "common/null_mask.h"
 #include "common/types/types.h"

--- a/src/storage/index/hash_index.cpp
+++ b/src/storage/index/hash_index.cpp
@@ -41,96 +41,6 @@ HashIndex<T>::HashIndex(const DBFileIDAndName& dbFileIDAndName, BMFileHandle* fi
     oSlots = diskArrays.getDiskArray<Slot<T>>(NUM_HASH_INDEXES + indexPos);
 }
 
-// For read transactions, local storage is skipped, lookups are performed on the persistent
-// storage. For write transactions, lookups are performed in the local storage first, then in
-// the persistent storage if necessary. In details, there are three cases for the local storage
-// lookup:
-// - the key is found in the local storage, directly return true;
-// - the key has been marked as deleted in the local storage, return false;
-// - the key is neither deleted nor found in the local storage, lookup in the persistent
-// storage.
-template<typename T>
-bool HashIndex<T>::lookupInternal(Transaction* transaction, Key key, offset_t& result) {
-    if (transaction->isReadOnly()) {
-        return lookupInPersistentIndex(transaction->getType(), key, result);
-    } else {
-        KU_ASSERT(transaction->isWriteTransaction());
-        auto localLookupState = localStorage->lookup(key, result);
-        if (localLookupState == HashIndexLocalLookupState::KEY_DELETED) {
-            return false;
-        } else if (localLookupState == HashIndexLocalLookupState::KEY_FOUND) {
-            return true;
-        } else {
-            KU_ASSERT(localLookupState == HashIndexLocalLookupState::KEY_NOT_EXIST);
-            return lookupInPersistentIndex(transaction->getType(), key, result);
-        }
-    }
-}
-
-// For deletions, we don't check if the deleted keys exist or not. Thus, we don't need to check
-// in the persistent storage and directly delete keys in the local storage.
-template<typename T>
-void HashIndex<T>::deleteInternal(Key key) const {
-    localStorage->deleteKey(key);
-}
-
-// For insertions, we first check in the local storage. There are three cases:
-// - the key is found in the local storage, return false;
-// - the key is marked as deleted in the local storage, insert the key to the local storage;
-// - the key doesn't exist in the local storage, check if the key exists in the persistent
-// index, if
-//   so, return false, else insert the key to the local storage.
-template<typename T>
-bool HashIndex<T>::insertInternal(Key key, offset_t value) {
-    offset_t tmpResult;
-    auto localLookupState = localStorage->lookup(key, tmpResult);
-    if (localLookupState == HashIndexLocalLookupState::KEY_FOUND) {
-        return false;
-    } else if (localLookupState == HashIndexLocalLookupState::KEY_NOT_EXIST) {
-        if (lookupInPersistentIndex(TransactionType::WRITE, key, tmpResult)) {
-            return false;
-        }
-    }
-    return localStorage->insert(key, value);
-}
-
-template<typename T>
-size_t HashIndex<T>::append(const IndexBuffer<BufferKeyType>& buffer) {
-    // Check if values already exist in persistent storage
-    if (indexHeaderForWriteTrx.numEntries > 0) {
-        common::offset_t result;
-        for (size_t i = 0; i < buffer.size(); i++) {
-            const auto& [key, value] = buffer[i];
-            if (lookupInPersistentIndex(transaction::TransactionType::WRITE, key, result)) {
-                return i;
-            }
-        }
-    }
-    return localStorage->append(buffer);
-}
-
-template<typename T>
-bool HashIndex<T>::lookupInPersistentIndex(TransactionType trxType, Key key, offset_t& result) {
-    auto& header = trxType == TransactionType::READ_ONLY ? this->indexHeaderForReadTrx :
-                                                           this->indexHeaderForWriteTrx;
-    // There may not be any primary key slots if we try to lookup on an empty index
-    if (header.numEntries == 0) {
-        return false;
-    }
-    auto hashValue = HashIndexUtils::hash(key);
-    auto fingerprint = HashIndexUtils::getFingerprintForHash(hashValue);
-    auto iter =
-        getSlotIterator(HashIndexUtils::getPrimarySlotIdForHash(header, hashValue), trxType);
-    do {
-        auto entryPos = findMatchedEntryInSlot(trxType, iter.slot, key, fingerprint);
-        if (entryPos != SlotHeader::INVALID_ENTRY_POS) {
-            result = iter.slot.entries[entryPos].value;
-            return true;
-        }
-    } while (nextChainedSlot(trxType, iter));
-    return false;
-}
-
 template<typename T>
 void HashIndex<T>::deleteFromPersistentIndex(Key key) {
     auto trxType = TransactionType::WRITE;
@@ -159,19 +69,6 @@ inline common::hash_t HashIndex<ku_string_t>::hashStored(transaction::Transactio
     auto str = overflowFileHandle->readString(TransactionType::WRITE, key);
     function::Hash::operation(str, hash);
     return hash;
-}
-
-template<typename T>
-entry_pos_t HashIndex<T>::findMatchedEntryInSlot(TransactionType trxType, const Slot<T>& slot,
-    Key key, uint8_t fingerprint) const {
-    for (auto entryPos = 0u; entryPos < getSlotCapacity<T>(); entryPos++) {
-        if (slot.header.isEntryValid(entryPos) &&
-            slot.header.fingerprints[entryPos] == fingerprint &&
-            equals(trxType, key, slot.entries[entryPos].key)) {
-            return entryPos;
-        }
-    }
-    return SlotHeader::INVALID_ENTRY_POS;
 }
 
 template<typename T>
@@ -223,16 +120,6 @@ bool HashIndex<T>::rollbackInMemory() {
     oSlots->rollbackInMemoryIfNecessary();
     localStorage->clear();
     return true;
-}
-
-template<>
-inline bool HashIndex<ku_string_t>::equals(transaction::TransactionType trxType,
-    std::string_view keyToLookup, const ku_string_t& keyInEntry) const {
-    if (HashIndexUtils::areStringPrefixAndLenEqual(keyToLookup, keyInEntry)) {
-        auto entryKeyString = overflowFileHandle->readString(trxType, keyInEntry);
-        return memcmp(keyToLookup.data(), entryKeyString.c_str(), entryKeyString.length()) == 0;
-    }
-    return false;
 }
 
 template<typename T>

--- a/src/storage/index/in_mem_hash_index.cpp
+++ b/src/storage/index/in_mem_hash_index.cpp
@@ -176,86 +176,6 @@ Slot<T>* InMemHashIndex<T>::clearNextOverflowAndAdvanceIter(SlotIterator& iter) 
 }
 
 template<typename T>
-size_t InMemHashIndex<T>::append(const IndexBuffer<BufferKeyType>& buffer) {
-    reserve(indexHeader.numEntries + buffer.size());
-    // Do both searches after splitting. Returning early if the key already exists isn't a
-    // particular concern and doing both after splitting allows the slotID to be reused
-    common::hash_t hashes[BUFFER_SIZE];
-    for (size_t i = 0; i < buffer.size(); i++) {
-        hashes[i] = HashIndexUtils::hash(buffer[i].first);
-    }
-    for (size_t i = 0; i < buffer.size(); i++) {
-        auto& [key, value] = buffer[i];
-        if (!appendInternal(key, value, hashes[i])) {
-            return i;
-        }
-    }
-    return buffer.size();
-}
-
-template<typename T>
-entry_pos_t InMemHashIndex<T>::findEntry(SlotIterator& iter, Key key, uint8_t fingerprint) {
-    do {
-        auto numEntries = iter.slot->header.numEntries();
-        KU_ASSERT(numEntries == std::countr_one(iter.slot->header.validityMask));
-        for (auto entryPos = 0u; entryPos < numEntries; entryPos++) {
-            if (iter.slot->header.fingerprints[entryPos] == fingerprint &&
-                equals(key, iter.slot->entries[entryPos].key)) [[unlikely]] {
-                // Value already exists
-                return entryPos;
-            }
-        }
-        if (numEntries < getSlotCapacity<T>()) {
-            return SlotHeader::INVALID_ENTRY_POS;
-        }
-    } while (nextChainedSlot(iter));
-    return SlotHeader::INVALID_ENTRY_POS;
-}
-
-template<typename T>
-bool InMemHashIndex<T>::appendInternal(Key key, common::offset_t value, common::hash_t hash) {
-    auto fingerprint = HashIndexUtils::getFingerprintForHash(hash);
-    auto slotID = HashIndexUtils::getPrimarySlotIdForHash(this->indexHeader, hash);
-    SlotIterator iter(slotID, this);
-    // The builder never keeps holes and doesn't support deletions
-    // Check the valid entries, then insert at the end if we don't find one which matches
-    auto entryPos = findEntry(iter, key, fingerprint);
-    auto numEntries = iter.slot->header.numEntries();
-    if (entryPos != SlotHeader::INVALID_ENTRY_POS) {
-        // The key already exists
-        return false;
-    } else if (numEntries < getSlotCapacity<T>()) [[likely]] {
-        // The key does not exist and the last slot has free space
-        insert(key, iter.slot, numEntries, value, fingerprint);
-        this->indexHeader.numEntries++;
-        return true;
-    }
-    // The last slot is full. Insert a new one
-    insertToNewOvfSlot(key, iter.slot, value, fingerprint);
-    this->indexHeader.numEntries++;
-    return true;
-}
-
-template<typename T>
-bool InMemHashIndex<T>::lookup(Key key, offset_t& result) {
-    // This needs to be fast if the builder is empty since this function is always tried
-    // when looking up in the persistent hash index
-    if (this->indexHeader.numEntries == 0) {
-        return false;
-    }
-    auto hashValue = HashIndexUtils::hash(key);
-    auto fingerprint = HashIndexUtils::getFingerprintForHash(hashValue);
-    auto slotId = HashIndexUtils::getPrimarySlotIdForHash(this->indexHeader, hashValue);
-    SlotIterator iter(slotId, this);
-    auto entryPos = findEntry(iter, key, fingerprint);
-    if (entryPos != SlotHeader::INVALID_ENTRY_POS) {
-        result = iter.slot->entries[entryPos].value;
-        return true;
-    }
-    return false;
-}
-
-template<typename T>
 uint32_t InMemHashIndex<T>::allocatePSlots(uint32_t numSlotsToAllocate) {
     auto oldNumSlots = pSlots->size();
     auto newNumSlots = oldNumSlots + numSlotsToAllocate;
@@ -291,24 +211,6 @@ Slot<T>* InMemHashIndex<T>::getSlot(const SlotInfo& slotInfo) const {
 }
 
 template<typename T>
-inline void InMemHashIndex<T>::insertToNewOvfSlot(Key key, Slot<T>* previousSlot,
-    common::offset_t offset, uint8_t fingerprint) {
-    auto newSlotId = allocateAOSlot();
-    previousSlot->header.nextOvfSlotId = newSlotId;
-    auto newSlot = getSlot(SlotInfo{newSlotId, SlotType::OVF});
-    auto entryPos = 0u; // Always insert to the first entry when there is a new slot.
-    insert(key, newSlot, entryPos, offset, fingerprint);
-}
-
-template<>
-void InMemHashIndex<ku_string_t>::insert(std::string_view key, Slot<ku_string_t>* slot,
-    uint8_t entryPos, offset_t offset, uint8_t fingerprint) {
-    auto& entry = slot->entries[entryPos];
-    entry = SlotEntry<ku_string_t>(overflowFileHandle->writeString(key), offset);
-    slot->header.setEntryValid(entryPos, fingerprint);
-}
-
-template<typename T>
 common::hash_t InMemHashIndex<T>::hashStored(const T& key) const {
     return HashIndexUtils::hash(key);
 }
@@ -318,69 +220,6 @@ common::hash_t InMemHashIndex<ku_string_t>::hashStored(const ku_string_t& key) c
     auto kuString = key;
     return HashIndexUtils::hash(
         overflowFileHandle->readString(transaction::TransactionType::WRITE, kuString));
-}
-
-template<>
-bool InMemHashIndex<ku_string_t>::equals(std::string_view keyToLookup,
-    const ku_string_t& keyInEntry) const {
-    // Checks if prefix and len matches first.
-    if (!HashIndexUtils::areStringPrefixAndLenEqual(keyToLookup, keyInEntry)) {
-        return false;
-    }
-    if (keyInEntry.len <= ku_string_t::PREFIX_LENGTH) {
-        // For strings shorter than PREFIX_LENGTH, the result must be true.
-        return true;
-    } else if (keyInEntry.len <= ku_string_t::SHORT_STR_LENGTH) {
-        // For short strings, whose lengths are larger than PREFIX_LENGTH, check if their actual
-        // values are equal.
-        return memcmp(keyToLookup.data(), keyInEntry.prefix, keyInEntry.len) == 0;
-    } else {
-        // For long strings, compare with overflow data
-        return overflowFileHandle->equals(transaction::TransactionType::WRITE, keyToLookup,
-            keyInEntry);
-    }
-}
-
-template<typename T>
-bool InMemHashIndex<T>::deleteKey(Key key) {
-    if (this->indexHeader.numEntries == 0) {
-        return false;
-    }
-    auto hashValue = HashIndexUtils::hash(key);
-    auto fingerprint = HashIndexUtils::getFingerprintForHash(hashValue);
-    auto slotId = HashIndexUtils::getPrimarySlotIdForHash(this->indexHeader, hashValue);
-    SlotIterator iter(slotId, this);
-    std::optional<entry_pos_t> deletedPos = 0;
-    do {
-        for (auto entryPos = 0u; entryPos < getSlotCapacity<T>(); entryPos++) {
-            if (iter.slot->header.isEntryValid(entryPos) &&
-                iter.slot->header.fingerprints[entryPos] == fingerprint &&
-                equals(key, iter.slot->entries[entryPos].key)) {
-                deletedPos = entryPos;
-                iter.slot->header.setEntryInvalid(entryPos);
-                break;
-            }
-        }
-        if (deletedPos.has_value()) {
-            break;
-        }
-    } while (nextChainedSlot(iter));
-
-    if (deletedPos.has_value()) {
-        // Find the last valid entry and move it into the deleted position
-        auto newIter = iter;
-        while (nextChainedSlot(newIter))
-            ;
-        if (newIter.slotInfo != iter.slotInfo ||
-            *deletedPos != newIter.slot->header.numEntries() - 1) {
-            auto lastEntryPos = newIter.slot->header.numEntries();
-            iter.slot->entries[*deletedPos] = newIter.slot->entries[lastEntryPos];
-            iter.slot->header.setEntryValid(*deletedPos,
-                newIter.slot->header.fingerprints[lastEntryPos]);
-            newIter.slot->header.setEntryInvalid(lastEntryPos);
-        }
-    }
-    return false;
 }
 
 template class InMemHashIndex<int64_t>;

